### ISSUE TITLE
TODO: Process deposits in order

### ISSUE THREAD

**Issue by ralexstokes** [ethereum/eth2.0-specs#389]
From the spec:

```
[TODO: add logic to ensure that deposits from 1.0 chain are processed in order]
```

Do we just mean to process `deposits` in a given block by the `deposit_data.timestamp`?

If so, I can make a PR w/ this change. Wasn't sure if there was something more involved here around making sure the entire mechanism works e.g. across updates to the `deposit_root`.

**Comment by djrtwo** [comment]
This would need to be a stronger validity condition than timestamp. We would likely either use the `previous_receipt_root` and force the deposits to be processed in a valid chain. Or we could add an index to the `Deposit` events from the pow chain and just process the indices in order.

**Comment by mkalinin** [comment]
There is an idea of merklizing `merkle_tree_index` or `deposit_count` value within the other `deposit_data` in the contract. This would allow to avoid querying a deposit contract which is needed to match indexes in deposit records of processing block. This option, also, requires to keep `latest_deposit_index` in the state.

**Comment by Nashatyrev** [comment]
I would also suggest in advance that if we are going to store `latest_deposit_index` in the `BeaconState` it would be implementation friendly to store `latest_deposit_eth1_data : Eth1Data` (either instead or together with the `latest_deposit_index`). 
That will help to avoid maintaining additional index for proposer to fetch next Deposits for block creation.

**Comment by djrtwo** [comment]
closed via #594
